File: swiftPrep/Sources/Graph/Prim.swift

import Tree

func prim(neighbors: [String: [String: Int]]) -> [(String, String, Int)] {
	
	guard let (startingVertex, _) = neighbors.first else { return [(String, String, Int)]() }
	var visited = Set([startingVertex])
	var priorityQ = Heap<(Int, String, String)>(orderedBy: <)
	
	var edges = [(String, String, Int)]()

	for (neighbor, edgeWeight) in neighbors[startingVertex]! {
		priorityQ.push((edgeWeight, startingVertex, neighbor))
	}

	while let (edgeWeight, u, v) = priorityQ.pop() {
		if !visited.contains(v) {
			visited.insert(v)

			for (w, nextEdgeWeight) in neighbors[v]! {
				if !visited.contains(w) {
					priorityQ.push((nextEdgeWeight, v, w))
				}
			}

			edges.append((u, v, edgeWeight))
		}
	}

	return edges
}

func minimumSpanningTree(neighbors: [String: [String: Int]]) -> Graph {
	let g = Graph(type: .undirected)

	for (u, v, edgeWeight) in prim(neighbors: neighbors) {
		g.addEdge(from: u, to: v, edgeWeight: edgeWeight)
	}

	return g
}

File: pythonPrep/sources/graph/prim.py
import heapq
from sources.graph.graph import *

def prim(graph):
	""" 
	Yield the minimum edges in the graph using Prim's Algorithm. 

	Time: O(|E| log |V|) using binary heap (faster with fibonacci heap)
	"""
	neighbors = graph.neighbors
	initial_vertex = next(iter(neighbors))
	visited = set([initial_vertex])
	priority_q = []

	for neighbor, edge_weight in neighbors[initial_vertex].items():
		heapq.heappush(priority_q, (edge_weight, initial_vertex, neighbor))

	while len(priority_q) > 0:
		edge_weight, u, v = heapq.heappop(priority_q)

		if v not in visited:
			visited.add(v)

			for w, next_edge_weight in neighbors[v].items():
				if w not in visited:
					heapq.heappush(priority_q, (next_edge_weight, v, w))

			yield u, v, edge_weight

def minimum_spanning_tree(graph):
	""" Return a minimum spanning tree for the graph. """
	ug = Graph(GraphType.UNDIRECTED)

	for u, v, edge_weight in prim(graph):
		ug.insert_edge(u, v, edge_weight)

	return ug
